File: cerebellum/Movement/robotic_arm_server.py
# Import necessary libraries
import time
import zmq
import json
from adafruit_servokit import ServoKit

# --- Servo Calibration (Use your found values) ---
MIN_PULSE = 500
MAX_PULSE = 2850

# --- ZMQ Configuration ---
COMMAND_PORT = "5555"  # Port for REQ/REP commands
PUBLISH_PORT = "5556"  # Port for PUB/SUB state publishing
PUB_TOPIC = "servo_state" # ZMQ topic for publishing

# --- Initialize ServoKit ---
kit = ServoKit(channels=16)

# --- Servo Definitions ---
# We use a dictionary to easily access servos by their ID
servos = {
    0: kit.servo[0],
    1: kit.servo[1],
    2: kit.servo[2],
    3: kit.servo[3],
    4: kit.servo[4],
}

# --- State Management ---
# Holds the last known angle for each servo.
# Initialize with your rest pose values.
current_angles = {
    0: 12,   # servo_A
    1: 90,   # servo_B
    2: 130,  # servo_C
    3: 100,  # servo_D
    4: 90    # servo_E
}

# --- Helper Functions (from your script) ---

def set_servo_calibration(servo_object):
    """Applies the calibrated pulse width to a servo object."""
    servo_object.actuation_range = 180
    servo_object.set_pulse_width_range(MIN_PULSE, MAX_PULSE)

def move_servo(servo_object, to_angle, from_angle=None, delay=0.035):
    """
    Moves a servo to a desired angle, with an optional smooth transition.
    Includes error handling for I2C communication failures.
    """
    if from_angle is not None:
        step = 1 if from_angle < to_angle else -1
        for angle in range(from_angle, to_angle + step, step):
            try:
                # This is the line that can fail due to hardware issues
                servo_object.angle = angle
            except OSError as e:
                # Catch the I/O error, print a warning, and continue running.
                print(f"!! WARNING: I2C communication error: {e}. Skipping step.")
                # 'continue' will skip this single angle and proceed with the loop.
                continue
            
            time.sleep(delay)
    else:
        try:
            # Also handle errors for direct (non-smooth) movements
            servo_object.angle = to_angle
        except OSError as e:
            print(f"!! WARNING: I2C communication error on direct move: {e}.")

    # We update the global state *after* the move is complete
    # Note: We'll find the servo_id to update the state dict
    for servo_id, obj in servos.items():
        if obj == servo_object:
            current_angles[servo_id] = to_angle
            break

def rest_pose():
    """Moves all servos to a predefined resting position."""
    print("Moving to rest pose...")
    # Use current_angles for smooth transitions to rest
    move_servo(servos[0], 12, from_angle=current_angles[0])
    move_servo(servos[1], 90, from_angle=current_angles[1])
    move_servo(servos[2], 130, from_angle=current_angles[2])
    move_servo(servos[3], 100, from_angle=current_angles[3])
    move_servo(servos[4], 90, from_angle=current_angles[4])
    print("In rest pose.")

# --- Main Server Function ---

def run_server():
    # Setup ZMQ context and sockets
    context = zmq.Context()
    
    # Socket for receiving commands
    command_socket = context.socket(zmq.REP)
    command_socket.bind(f"tcp://*:{COMMAND_PORT}")
    
    # Socket for publishing state
    publish_socket = context.socket(zmq.PUB)
    publish_socket.bind(f"tcp://*:{PUBLISH_PORT}")
    
    print(f"Servo server started...")
    print(f"Listening for commands on port {COMMAND_PORT}")
    print(f"Publishing state on port {PUBLISH_PORT}")

    # Apply calibration to all servos
    print("Calibrating servos...")
    for servo in servos.values():
        set_servo_calibration(servo)
    
    # Go to initial rest pose
    rest_pose()
    
    try:
        while True:
            # 1. Wait for a command
            message = command_socket.recv_string()
            print(f"Received command: '{message}'")
            
            response = "Acknowledged. No action taken."
            execute_move = False
            
            # 2. Parse the command
            try:
                parts = message.split()
                if len(parts) == 4 and parts[0] == 'servo' and parts[2] == 'angle':
                    servo_id = int(parts[1])
                    angle = int(parts[3])
                    
                    # 3. Validate the command
                    if servo_id not in servos:
                        response = f"Error: Invalid servo ID '{servo_id}'."
                    elif not (0 <= angle <= 180):
                        response = f"Error: Invalid angle '{angle}'. Must be 0-180."
                    else:
                        # Command is valid
                        response = f"OK: Moving servo {servo_id} to {angle}."
                        execute_move = True
                
                elif len(parts) == 1 and parts[0] == 'rest':
                    rest_pose()
                    response = "OK: Moved to rest pose."
                    execute_move = False
                
                else:
                    response = "Error: Invalid command format. Use 'servo X angle Y' or 'rest'."

            except (ValueError, IndexError):
                response = "Error: Invalid command. Could not parse."
            
            # 4. Execute the command (if valid)
            if execute_move:
                servo_obj = servos[servo_id]
                
                # This function now contains the error handling
                move_servo(servo_obj, angle)
                # Update the state tracker
                current_angles[servo_id] = angle
            # 5. Send the reply
            command_socket.send_string(response)
            
            # 6. Publish the new state (always, even on error)
            state_message = json.dumps(current_angles)
            publish_socket.send_string(f"{PUB_TOPIC} {state_message}")
            print(f"Published state: {state_message}")

    except KeyboardInterrupt:
        print("\nShutting down server...")
        rest_pose() # Go to rest on shutdown
    
    finally:
        # Clean up
        command_socket.close()
        publish_socket.close()
        context.term()
        print("Server shut down.")

# --- Start the server ---
if __name__ == "__main__":
    run_server()

File: cerebellum/Movement/robotic_arm_server.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <map>
#include <chrono>
#include <thread>
#include <cmath>
#include <zmq.hpp>
#include <nlohmann/json.hpp>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>

using json = nlohmann::json;

// --- Config ---
const std::string COMMAND_PORT = "5555";
const std::string PUBLISH_PORT = "5556";
const std::string PUB_TOPIC = "servo_state";

const int PCA9685_ADDRESS = 0x40; // Default address
const int MODE1 = 0x00;
const int PRESCALE = 0xFE;
const int LED0_ON_L = 0x06;

// Pulse width bounds matching Python
const int MIN_PULSE = 500;
const int MAX_PULSE = 2850;

// State management
std::map<int, int> current_angles = {
    {0, 12},   // servo_A
    {1, 90},   // servo_B
    {2, 130},  // servo_C
    {3, 100},  // servo_D
    {4, 90}    // servo_E
};

int i2c_file;

// I2C Helpers
void write_byte(int reg, int value) {
    unsigned char buf[2] = {(unsigned char)reg, (unsigned char)value};
    if (write(i2c_file, buf, 2) != 2) {
        std::cerr << "!! WARNING: I2C write failed for reg " << reg << std::endl;
    }
}

int read_byte(int reg) {
    unsigned char buf[1] = {(unsigned char)reg};
    if (write(i2c_file, buf, 1) != 1) {
        return -1;
    }
    unsigned char val;
    if (read(i2c_file, &val, 1) != 1) {
        return -1;
    }
    return val;
}

// PCA9685 initialization
void init_pca9685() {
    std::string i2c_filename = "/dev/i2c-1";
    i2c_file = open(i2c_filename.c_str(), O_RDWR);
    if (i2c_file < 0) {
        std::cerr << "Could not open I2C bus " << i2c_filename << ". Will continue in mock mode." << std::endl;
    } else {
        if (ioctl(i2c_file, I2C_SLAVE, PCA9685_ADDRESS) < 0) {
            std::cerr << "Failed to acquire bus access and/or talk to slave." << std::endl;
        } else {
            write_byte(MODE1, 0x00);

            // Set frequency to 50Hz
            float freq = 50.0;
            float prescaleval = 25000000.0;
            prescaleval /= 4096.0;
            prescaleval /= freq;
            prescaleval -= 1.0;
            int prescale = floor(prescaleval + 0.5);

            int oldmode = read_byte(MODE1);
            int newmode = (oldmode & 0x7F) | 0x10; // sleep
            write_byte(MODE1, newmode);
            write_byte(PRESCALE, prescale);
            write_byte(MODE1, oldmode);
            std::this_thread::sleep_for(std::chrono::milliseconds(5));
            write_byte(MODE1, oldmode | 0xa0);
        }
    }
}

// Convert angle (0-180) to PCA9685 tick value (0-4095)
int angle_to_tick(int angle) {
    // Pulse width in microseconds
    float pulse_width = MIN_PULSE + (float)angle / 180.0 * (MAX_PULSE - MIN_PULSE);
    // Period for 50Hz is 20ms = 20000us. Ticks = 4096 * (pulse_width / 20000)
    int tick = (int)(4096.0 * (pulse_width / 20000.0));
    return tick;
}

void set_pwm(int channel, int on, int off) {
    if (i2c_file >= 0) {
        write_byte(LED0_ON_L + 4 * channel, on & 0xFF);
        write_byte(LED0_ON_L + 4 * channel + 1, on >> 8);
        write_byte(LED0_ON_L + 4 * channel + 2, off & 0xFF);
        write_byte(LED0_ON_L + 4 * channel + 3, off >> 8);
    }
}

void move_servo(int servo_id, int to_angle, bool smooth = false) {
    if (current_angles.find(servo_id) == current_angles.end()) {
        std::cerr << "Unknown servo id " << servo_id << std::endl;
        return;
    }

    if (smooth) {
        int from_angle = current_angles[servo_id];
        int step = (from_angle < to_angle) ? 1 : -1;
        for (int angle = from_angle; angle != to_angle + step; angle += step) {
            set_pwm(servo_id, 0, angle_to_tick(angle));
            std::this_thread::sleep_for(std::chrono::milliseconds(35));
        }
    } else {
        set_pwm(servo_id, 0, angle_to_tick(to_angle));
    }

    current_angles[servo_id] = to_angle;
}

void rest_pose() {
    std::cout << "Moving to rest pose..." << std::endl;
    move_servo(0, 12, true);
    move_servo(1, 90, true);
    move_servo(2, 130, true);
    move_servo(3, 100, true);
    move_servo(4, 90, true);
    std::cout << "In rest pose." << std::endl;
}

void calibration_dance() {
    std::cout << "Starting Calibration Dance..." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(2));
    move_servo(0, 45, false);
    std::this_thread::sleep_for(std::chrono::seconds(1));
    move_servo(0, 135, false);
    std::this_thread::sleep_for(std::chrono::seconds(1));
    move_servo(0, 90, false);

    move_servo(4, 0, false);
    std::this_thread::sleep_for(std::chrono::seconds(1));
    move_servo(4, 180, false);
    std::this_thread::sleep_for(std::chrono::seconds(1));
    move_servo(4, 90, false);
    std::cout << "Calibration Dance complete." << std::endl;
}

std::string process_command(const std::string& message) {
    std::istringstream iss(message);
    std::vector<std::string> parts;
    std::string part;
    while (iss >> part) {
        parts.push_back(part);
    }

    if (parts.empty()) return "Error: Empty command.";

    if (parts.size() == 4 && parts[0] == "servo" && parts[2] == "angle") {
        try {
            int servo_id = std::stoi(parts[1]);
            int angle = std::stoi(parts[3]);

            if (current_angles.find(servo_id) == current_angles.end()) {
                return "Error: Invalid servo ID.";
            }
            if (angle < 0 || angle > 180) {
                return "Error: Invalid angle. Must be 0-180.";
            }

            move_servo(servo_id, angle, false);
            return "OK: Moving servo " + std::to_string(servo_id) + " to " + std::to_string(angle) + ".";
        } catch (...) {
            return "Error: Could not parse integers.";
        }
    } else if (parts.size() == 1 && parts[0] == "rest") {
        rest_pose();
        return "OK: Moved to rest pose.";
    } else if (parts.size() == 1 && parts[0] == "calibrate") {
        calibration_dance();
        return "OK: Calibration Dance complete.";
    }

    return "Error: Invalid command format. Use 'servo X angle Y', 'rest', or 'calibrate'.";
}

int main() {
    init_pca9685();

    zmq::context_t context(1);
    zmq::socket_t command_socket(context, ZMQ_REP);
    command_socket.bind("tcp://*:" + COMMAND_PORT);

    zmq::socket_t publish_socket(context, ZMQ_PUB);
    publish_socket.bind("tcp://*:" + PUBLISH_PORT);

    std::cout << "Servo server started..." << std::endl;
    std::cout << "Listening for commands on port " << COMMAND_PORT << std::endl;
    std::cout << "Publishing state on port " << PUBLISH_PORT << std::endl;

    rest_pose();

    while (true) {
        zmq::message_t request;
        auto res = command_socket.recv(request, zmq::recv_flags::none);
        if (res) {
            std::string message(static_cast<char*>(request.data()), request.size());
            std::cout << "Received command: '" << message << "'" << std::endl;

            std::string response = process_command(message);

            zmq::message_t reply(response.size());
            memcpy(reply.data(), response.c_str(), response.size());
            command_socket.send(reply, zmq::send_flags::none);

            // Publish state
            json state_json;
            // The go script expects keys as strings
            for (const auto& pair : current_angles) {
                state_json[std::to_string(pair.first)] = pair.second;
            }

            std::string state_message = PUB_TOPIC + " " + state_json.dump();
            zmq::message_t pub_msg(state_message.size());
            memcpy(pub_msg.data(), state_message.c_str(), state_message.size());
            publish_socket.send(pub_msg, zmq::send_flags::none);

            std::cout << "Published state: " << state_json.dump() << std::endl;
        }
    }

    return 0;
}
